### PR TITLE
mobile: Stop the DNS cache when the network is unavailable

### PR DESCRIPTION
**Pull request by fredyw**
This PR fixes an issue on Android when the network was unavailable, Envoy Mobile would always refresh the DNS cache causing spurious DNS errors. This PR fixes the issue by cancelling the pending DNS queries and stopping the DNS refresh and timeout timers when the network is unavailable. As a result, this requires adding a new `DnsCache::stop` function.

The `InternalEngine` API has also been updated by introducing `onDefaultNetworkAvailable`, `onDefaultNetworkChanged` and `onDefaultNetworkUnavailable` callbacks instead of using `setPreferredNetwork`, a method that has been incorrectly used for on network changed callback. By having these new callbacks, we now have a fine-grained control to refresh and stop the DNS caches.

The `AndroidNetworkMonitor` has been reimplemented to use [ConnectivityManager.Network](https://developer.android.com/reference/android/net/ConnectivityManager.NetworkCallback) instead of the deprecated API and more tests has also been added.

Risk Level: medium
Testing: unit and integration tests
Docs Changes: n/a
Release Notes: n/a
Platform Specific Features: dns, mobile

### CODE REVIEW

File: mobile/library/common/internal_engine.h [hunk 1]
```diff
@@ -106,18 +106,27 @@ class InternalEngine : public Logger::Loggable<Logger::Id::main> {
...
    *
-   * - Sets the preferred network.
    * - Check for IPv6 connectivity. If there is no IPv6 no connectivity, it will call
    *   `setIpVersionToRemove` in the DNS cache implementation to remove the IPv6 addresses from
    *   the DNS response in the subsequent DNS resolutions.
    * - Force refresh the hosts in the DNS cache (will take `setIpVersionToRemove` into account).
    * - Optionally (if configured) clear HTTP/3 broken status.
    */
-  envoy_status_t setPreferredNetwork(NetworkType network);
+  void onNetworkAvailable();
```
Comment (RyanTheOptimist) [inline mobile/library/common/internal_engine.h:121]: I believe In the Chrome APIs, there is a distinction between a network being made available and being made default. Do we want to do all of these things when a new network is available, or just when it's made default?

File: mobile/library/common/internal_engine.h [hunk 2]
```diff
@@ -106,18 +106,27 @@ class InternalEngine : public Logger::Loggable<Logger::Id::main> {
...
+
+  /**
+   * This functions does the following on network unavailable.
+   *
+   * - Cancel the DNS pending queries.
+   * - Stop the DNS timeout and refresh timers.
+   */
+  void onNetworkUnavailable();
```
Comment (RyanTheOptimist) [inline mobile/library/common/internal_engine.h:129]: Does this happen when any network is made unavailable, or only when the default network is?

**Review by RyanTheOptimist — COMMENTED** [review]
Two API questions...

File: source/extensions/common/dynamic_forward_proxy/dns_cache_impl.cc [hunk 3]
```diff
@@ -337,6 +337,27 @@ void DnsCacheImpl::setIpVersionToRemove(absl::optional<Network::Address::IpVersi
...
+  // transition and parameters may have changed.
+  resolver_->resetNetworking();
+
+  absl::ReaderMutexLock reader_lock{&primary_hosts_lock_};
+  for (auto& primary_host : primary_hosts_) {
+    if (primary_host.second->active_query_ != nullptr) {
+      primary_host.second->active_query_->cancel(
+          Network::ActiveDnsQuery::CancelReason::QueryAbandoned);
+      primary_host.second->active_query_ = nullptr;
+      primary_host.second->timeout_timer_->disableTimer();
```
Comment (abeyad) [inline source/extensions/common/dynamic_forward_proxy/dns_cache_impl.cc:352]: should this be done outside the `if`?  i.e. on L354?

File: mobile/library/java/io/envoyproxy/envoymobile/engine/AndroidNetworkMonitor.java [hunk 4]
```diff
@@ -73,42 +70,36 @@ private AndroidNetworkMonitor(Context context, EnvoyEngine envoyEngine) {
...
-      public void onCapabilitiesChanged(Network network, NetworkCapabilities networkCapabilities) {
-        handleNetworkChange();
+      public void onAvailable(@NonNull Network network) {
+        envoyEngine.onNetworkAvailable();
       }
+
       @Override
-      public void onLosing(Network network, int maxMsToLive) {
-        handleNetworkChange();
+      public void onCapabilitiesChanged(@NonNull Network network,
```
Comment (abeyad) [inline mobile/library/java/io/envoyproxy/envoymobile/engine/AndroidNetworkMonitor.java:86]: is this guaranteed to be invoked only when there was a change?  do we want to check that engine's current preferred network value before calling `setPreferredNetwork`?

**Review by abeyad — APPROVED** [review]
mobile LGTM, thanks for working on it!

**Review by RyanTheOptimist — APPROVED** [review]

**Pull request merged**